File: LeetCode/89-gray-code/Solution.java
import java.util.ArrayList;
import java.util.List;

class Solution {
  public List<Integer> grayCode(int n) {
    List<Integer> result = new ArrayList<>();
    result.add(0); // Initialize sequence with 0

    for (int i = 1; i <= n; i++) {
      int mask = 1 << (i - 1); // Set the (i-1)th bit using bit shift

      // Reverse iterate over current sequence and apply mask
      for (int j = result.size() - 1; j >= 0; j--) {
        result.add(result.get(j) | mask); // Add prefix 1 to mirrored numbers
      }
    }

  }
}

File: LeetCode/89-gray-code/solution.ts
function grayCodeIterative(n: number): number[] {
  const result: number[] = [0]; // Initialize with 0

  for (let i = 0; i < n; i++) {
    const mask = 1 << i; // Define mask for bit shifting

    // Reverse iterate and apply bitwise OR to create mirrored sequence
    for (let j = result.length - 1; j >= 0; j--) {
      result.push(result[j] | mask);
    }
  }

  return result;
}